src/core/server/saved_objects/migrationsv2/actions/calculate_exclude_filters.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { estypes } from '@elastic/elasticsearch';
+import * as Either from 'fp-ts/lib/Either';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import { RetryableEsClientError } from '.';
+import { ElasticsearchClient } from '../../../elasticsearch';
+import { SavedObjectTypeExcludeFromUpgradeFilterHook } from '../../types';
+import { catchRetryableEsClientErrors } from './catch_retryable_es_client_errors';
+
+export interface CalculateExcludeFiltersParams {
+  client: ElasticsearchClient;
+  excludeFromUpgradeFilterHooks: SavedObjectTypeExcludeFromUpgradeFilterHook[];
+}
+
+export interface CalculatedExcludeFilter {
+  /** Composite filter of all calculated filters */
+  excludeFilter: estypes.QueryDslQueryContainer;
+  /** Any errors that were encountered during filter calculation */
+  errors: any[];
+}
+
+export const calculateExcludeFilters = ({
+  client,
+  excludeFromUpgradeFilterHooks,
+}: CalculateExcludeFiltersParams): TaskEither.TaskEither<
+  RetryableEsClientError,
+  CalculatedExcludeFilter
+> => () => {
+  return Promise.all(
+    excludeFromUpgradeFilterHooks.map((hook) =>
```

Comment:
It contradicts a bit the migration from async to sync plugin hooks. Should we add a timeout for these `hooks` to be executed?

Reply:
> Is it absolutely necessary for `tasks` plugin to query the index?

Unfortunately, yes.

> Should we add a timeout for these hooks to be executed?
 
Good idea, I'll add that.

Reply:
> Unfortunately, yes.

Why? Cannot we just remove all the `failed` tasks instead of searching for the oldest one?

Reply:
For the `action_task_params` documents in the `.kibana` index we can't remove any that haven't been used by their associated tasks yet and this information is not available in the `.kibana` index.

Reply:
Unsure about the failed `tasks`, maybe @mikecote could answer?

Reply:
For the `tasks`, I was talking with @chrisronline , there's not really any task with `status:failed` that we need to keep, so removing the time filter in that query would be ok (the customer can always refer to the previous index for those as a backup).